### ISSUE TITLE
Confd is sub-matching applications leading to non-existing application

### ISSUE THREAD

**Issue by kalbasit** [deis/deis#4346]
In [nginx.conf:345](https://github.com/deis/deis/blob/v1.9.0/router/image/templates/nginx.conf#L345) `{{ range $kapp := ls $k8appdir }}` is sub-matching applications. For example I have two applications, one named `app` and one named `app-dev`, `$kapp` is being assigned to `api`, `api-dev` and `-dev`

debugging info:

```
core@ip-10-21-1-236 ~ $ etcdctl ls /registry/services/specs/
/registry/services/specs/default
/registry/services/specs/api
/registry/services/specs/goto
/registry/services/specs/eve-dev
/registry/services/specs/console-dev
/registry/services/specs/api-dev
```

`nse` into a `deis-router` container and changed `/app/templates/nginx.conf` between `#start k8s apps` and `# start builder` to match

```
    #start k8s apps
    {{ range $k8namespace := lsdir "/registry/services/specs/" }}
    {{ $k8appdir := printf "/registry/services/specs/%s" $k8namespace }}

    # $k8namespace = {{ $k8namespace }}
    # $k8appdir = {{ $k8appdir }}

    {{ range $kapp := ls $k8appdir }}
    {{ $k8appPath := printf "/registry/services/specs/%s/%s" $k8namespace $kapp}}
    {{ $k8Svc := getv $k8appPath }}

    # $kapp = {{ $kapp }}
    # $k8appPath = {{ $k8appPath }}
    # $k8Svc = {{ $k8Svc }}

    {{end}}{{end}}
}

## start builder
```

and ran `confd --node 10.21.1.236:4001 --confdir /app --log-level debug` here's the relevant section of `/opt/nginx/conf/nginx.conf`

```
    #start k8s apps



    # $k8namespace = api
    # $k8appdir = /registry/services/specs/api





    # $kapp = -dev
    # $k8appPath = /registry/services/specs/api/-dev
    # $k8Svc =





    # $kapp = app-29154
    # $k8appPath = /registry/services/specs/api/app-29154
    # $k8Svc = {"kind":"Service","apiVersion":"v1","metadata":{"name":"app-29154","namespace":"api","uid":"8f93c753-453b-11e5-badd-0eb4b14dac9f","creationTimestamp":"2015-08-17T23:56:21Z","labels":{"name":"api"}},"spec":{"ports":[{"protocol":"TCP","port":80,"targetPort":8091,"nodePort":0}],"selector":{"name":"api","type":"cmd"},"clusterIP":"10.100.109.66","type":"ClusterIP","sessionAffinity":"None"},"status":{"loadBalancer":{}}}




    # $k8namespace = api-dev
    # $k8appdir = /registry/services/specs/api-dev





    # $kapp = app-66429
    # $k8appPath = /registry/services/specs/api-dev/app-66429
    # $k8Svc = {"kind":"Service","apiVersion":"v1","metadata":{"name":"app-66429","namespace":"api-dev","uid":"4f299e1e-4c3b-11e5-83da-0eb4b14dac9f","creationTimestamp":"2015-08-26T21:42:11Z","labels":{"name":"api-dev"}},"spec":{"ports":[{"protocol":"TCP","port":80,"targetPort":8091,"nodePort":0}],"selector":{"name":"api-dev","type":"cmd"},"clusterIP":"10.100.78.1","type":"ClusterIP","sessionAffinity":"None"},"status":{"loadBalancer":{}}}




    # $k8namespace = console-dev
    # $k8appdir = /registry/services/specs/console-dev





    # $kapp = app-61568
    # $k8appPath = /registry/services/specs/console-dev/app-61568
    # $k8Svc = {"kind":"Service","apiVersion":"v1","metadata":{"name":"app-61568","namespace":"console-dev","uid":"3f1e2e66-4b82-11e5-83da-0eb4b14dac9f","creationTimestamp":"2015-08-25T23:37:28Z","labels":{"name":"console-dev"}},"spec":{"ports":[{"protocol":"TCP","port":80,"targetPort":80,"nodePort":0}],"selector":{"name":"console-dev","type":"cmd"},"clusterIP":"10.100.101.37","type":"ClusterIP","sessionAffinity":"None"},"status":{"loadBalancer":{}}}




    # $k8namespace = default
    # $k8appdir = /registry/services/specs/default





    # $kapp = kubernetes
    # $k8appPath = /registry/services/specs/default/kubernetes
    # $k8Svc = {"kind":"Service","apiVersion":"v1","metadata":{"name":"kubernetes","namespace":"default","uid":"8e3418ac-4539-11e5-badd-0eb4b14dac9f","creationTimestamp":"2015-08-17T23:42:00Z","labels":{"component":"apiserver","provider":"kubernetes"}},"spec":{"ports":[{"protocol":"TCP","port":443,"targetPort":443,"nodePort":0}],"clusterIP":"10.100.0.1","type":"ClusterIP","sessionAffinity":"None"},"status":{"loadBalancer":{}}}




    # $k8namespace = eve-dev
    # $k8appdir = /registry/services/specs/eve-dev





    # $kapp = app-81694
    # $k8appPath = /registry/services/specs/eve-dev/app-81694
    # $k8Svc = {"kind":"Service","apiVersion":"v1","metadata":{"name":"app-81694","namespace":"eve-dev","uid":"15255f38-468e-11e5-badd-0eb4b14dac9f","creationTimestamp":"2015-08-19T16:19:35Z","labels":{"name":"eve-dev"}},"spec":{"ports":[{"protocol":"TCP","port":80,"targetPort":80,"nodePort":0}],"selector":{"name":"eve-dev","type":"cmd"},"clusterIP":"10.100.115.21","type":"ClusterIP","sessionAffinity":"None"},"status":{"loadBalancer":{}}}




    # $k8namespace = goto
    # $k8appdir = /registry/services/specs/goto





    # $kapp = app-54117
    # $k8appPath = /registry/services/specs/goto/app-54117
    # $k8Svc = {"kind":"Service","apiVersion":"v1","metadata":{"name":"app-54117","namespace":"goto","uid":"c9be2e51-453f-11e5-badd-0eb4b14dac9f","creationTimestamp":"2015-08-18T00:26:37Z","labels":{"name":"goto"}},"spec":{"ports":[{"protocol":"TCP","port":80,"targetPort":8181,"nodePort":0}],"selector":{"name":"goto","type":"cmd"},"clusterIP":"10.100.61.65","type":"ClusterIP","sessionAffinity":"None"},"status":{"loadBalancer":{}}}
```

**Comment by aledbf** [comment]
@kalbasit the error is here https://github.com/deis/memkv/pull/1

**Comment by bacongobbler** [comment]
That looks like it's not just specific to kubernetes but with all scheulers. I'll take a look into it. Thanks @aledbf for the sleuthing :)

**Comment by carmstrong** [comment]
We need to get to this. @bacongobbler /cc @deis/production-team

**Comment by bacongobbler** [comment]
I wasn't able to reproduce this upstream:

```
><> git diff
diff --git a/src/github.com/kelseyhightower/confd/resource/template/template_test.go b/src/github.com/kelseyhightower/confd/resource/template/template_test.go
index 62642af..c9e5383 100644
--- a/src/github.com/kelseyhightower/confd/resource/template/template_test.go
+++ b/src/github.com/kelseyhightower/confd/resource/template/template_test.go
@@ -287,9 +287,7 @@ num: 3
 src = "test.conf.tmpl"
 dest = "./tmp/test.conf"
 keys = [
-    "/test/data/abc",
-    "/test/data/def",
-    "/test/data/ghi",
+    "/test/data",
 ]
 `,
                tmpl: `
@@ -305,11 +303,14 @@ value: def
 
 value: ghi
 
+value: ghi-jkl
+
 `,
                updateStore: func(tr *TemplateResource) {
                        tr.store.Set("/test/data/abc", "123")
                        tr.store.Set("/test/data/def", "456")
                        tr.store.Set("/test/data/ghi", "789")
+                       tr.store.Set("/test/data/ghi-jkl", "012")
                },
        },
```

I was able to get `./test` to work without issue. Could you please try reproducing this issue upstream? https://github.com/kelseyhightower/confd

**Comment by aledbf** [comment]
@bacongobbler the problem is that `deis-router` is not using the upstream version of `confd` https://github.com/deis/deis/blob/master/router/Dockerfile#L15

**Comment by bacongobbler** [comment]
@aledbf I was also unable to reproduce on https://github.com/deis/confd as well, which is what that is built from:

```
><> git diff
diff --git a/resource/template/template_test.go b/resource/template/template_test.go
index 066ebf5..1758c94 100644
--- a/resource/template/template_test.go
+++ b/resource/template/template_test.go
@@ -241,9 +241,7 @@ num: 3
 src = "test.conf.tmpl"
 dest = "./tmp/test.conf"
 keys = [
-    "/test/data/abc",
-    "/test/data/def",
-    "/test/data/ghi",
+    "/test/data",
 ]
 `,
                tmpl: `
@@ -259,11 +257,14 @@ value: def
 
 value: ghi
 
+value: ghi-jkl
+
 `,
                updateStore: func(tr *TemplateResource) {
                        tr.store.Set("/test/data/abc", "123")
                        tr.store.Set("/test/data/def", "456")
                        tr.store.Set("/test/data/ghi", "789")
+                       tr.store.Set("/test/data/ghi-jkl", "012")
                },
        },
```

I saw your patch (deis/memkv#1) but the patch bypasses a key check which might introduce further issues. That's when I tried going back upstream to see if the issue's found there but I have been having no luck with it.

**Comment by carmstrong** [comment]
@vilinios proposed a solution for this in #4753 - rather than patch confd upstream, if we could fix this on our end with a simple change, it'd be much preferred.

**Comment by carmstrong** [comment]
closed by #4753
